tasks/preprocess_upload.py

```diff
@@ -62,7 +64,10 @@ def run_impl(
             with redis_connection.lock(
                 lock_name,
                 timeout=60 * 5,
-                blocking_timeout=None,
+                # This is the timeout that this task will wait to wait for the lock. This should
+                # be non-zero as otherwise it waits indefinitely to get the lock, and ideally smaller than
+                # the blocking timeout for the upload task so this one goes first, although it can go second
```

Comment:
I'm kinda realizing now that the current setup could lead us into trouble...

Consider this scenario:
1 Preprocess and 1 Upload tasks execute at roughly the same time. PreProcess get's the lock. It can stay with the lock for a maximum of 5minutes. Let's say it keeps the lock for 5 minutes.

Upload waits 5s, doesn't get the lock. It retries (after [20s](https://github.com/codecov/worker/blob/main/tasks/upload.py#L337)). Then it waits 5 more seconds to get the lock and fails. It will retry again (after 40s), and wait 5 more seconds. Then it will [not retry again](https://github.com/codecov/worker/blob/main/tasks/upload.py#L324).
In total the task waiter 15s + 20s + 40s = 1m15s

In this scenario no Upload tasks would run for the commit and nothing would be processed

Reply:
That makes sense, so if we just ensure the upload task waits at least 5 minutes in the worst case scenario, that would fix this right?

Every unsuccessful run goes through blocking_timeout (0) + retry_countdown (20 * 2**self.request.retries), so it goes like 

0 retries: 5 + 20 = 25
1 retries: 5 + 40 = 45 (total 70)
2 retries: 5 + 20*4 = 85 (total 155)
3 retries: 5 + 20*8 = 165 (total 320)

So if we'd allow this to retry at most 3 times, this could solve this, or we could tweak the specific values of retrying. That's one way. 

Another is to shorten the time during preprocess and fit the upload retries around that. 

There's the elimination of the preprocess task altogether, which I can ask other people/engs but I suspect that will be met with more friction + unforeseen things cause we're suddenly unsupporting a command. 

Let me know what you think about this

Reply:
I would push for the elimination of the preprocess task altogether. It would be a good idea to get other opinions from the team, definetely.

It is not true that "we're suddenly unsupporting a command" if we drop that task. The api would still create an empty report.

In the meantime we can go with the proposal in the other PR to alleviate customer's problems (although for the record I don't particularly like that solution, but that's just 1 opinion)